artiq/gateware/shuttler.py

```diff
@@ -18,9 +18,80 @@
 
 from migen import *
 from misoc.interconnect.stream import Endpoint
+from misoc.interconnect.csr import *
 from misoc.cores.cordic import Cordic
+from migen.genlib.cdc import MultiReg
 from artiq.gateware.rtio import rtlink
 
+class Dac_Interface(Module, AutoCSR):
+    def __init__(self, pads, bit_width=14):
+        self.data = [[Signal(bit_width) for _ in range(2)] for _ in range(8)]
+        dac_clk_en = Signal(8)
+        self.clk_en = CSRStorage(8)
+        self.ready = CSRStorage()
+
+        self.phase_shift = CSR()
+        self.phase_shift_done = CSRStatus(reset=1)
+
+        mmcm_ps_fb = Signal()
+        mmcm_ps_output = Signal()
+        mmcm_ps_psdone = Signal()
+
+        # Generate DAC DDR CLK
+        # 125MHz to 125MHz with controllable phase shift,
+        # VCO @ 1000MHz.
+        # Phase is shifted by 45 degree by default
+        self.specials += \
+            Instance("MMCME2_ADV",
+                p_CLKIN1_PERIOD=8.0,
+                i_CLKIN1=ClockSignal(),
+                i_RST=ResetSignal(),
+                i_CLKINSEL=1, 
+
+                p_CLKFBOUT_MULT_F=8.0,
+                p_CLKOUT0_DIVIDE_F=8.0,
+                p_DIVCLK_DIVIDE=1,
+                p_CLKOUT0_PHASE=45.0,
+
+                o_CLKFBOUT=mmcm_ps_fb, i_CLKFBIN=mmcm_ps_fb,
+
+                p_CLKOUT0_USE_FINE_PS="TRUE",
+                o_CLKOUT0=mmcm_ps_output,
+
+                i_PSCLK=ClockSignal(),
+                i_PSEN=self.phase_shift.re,
+                i_PSINCDEC=self.phase_shift.r,
+                o_PSDONE=mmcm_ps_psdone,
+            )
+
+        self.sync += [
+            If(self.phase_shift.re, self.phase_shift_done.status.eq(0)),
+            If(mmcm_ps_psdone, self.phase_shift_done.status.eq(1))
+        ]
+
+        self.clock_domains.cd_dac_ddr = ClockDomain()
+        self.specials += [
+            Instance("BUFG", i_I=mmcm_ps_output, o_O=self.cd_dac_ddr.clk),
```

Comment:
> I think you can use a local clock buffer (BUFH, BUFIO and the like) since it just needs to drive a few ODDR.

I found that ODDR clock output pins are scattered across multiple clock regions/IO banks. I cannot directly connect the MMCM output to the ODDR Clock input port with other types of clock buffer.

I think using BUFG should have the cleanest code. Otherwise, I will need to create a clock distribution network of clock buffers to drive the OLOGIC inside different multiple clock region/io bank.